pkg/proxy/proxymap.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/bpf"
+)
+
+type Proxy4Key struct {
+	SAddr   types.IPv4
+	DPort   uint16
+	SPort   uint16
+	Nexthdr uint8
+}
+
+type Proxy4Value struct {
+	OrigDAddr types.IPv4
+	OrigDPort uint16
+	Lifetime  uint16
+}
+
+var (
+	proxy4Map = bpf.NewMap("cilium_proxy4",
+		bpf.MapTypeHash,
+		int(unsafe.Sizeof(Proxy4Key{})),
+		int(unsafe.Sizeof(Proxy4Value{})),
+		8192)
+)
+
+func (k Proxy4Key) NewValue() bpf.MapValue {
+	return &Proxy4Value{}
+}
+
+func (k *Proxy4Key) GetKeyPtr() unsafe.Pointer {
+	return unsafe.Pointer(k)
+}
+
+func (k *Proxy4Key) String() string {
+	return fmt.Sprintf("%s:%d (%d) => %d", k.SAddr.IP().String(), k.SPort, k.Nexthdr, k.DPort)
+}
+
+func (k *Proxy4Key) Convert() *Proxy4Key {
+	n := *k
+	n.SPort = common.Swab16(n.SPort)
+	n.DPort = common.Swab16(n.DPort)
+	return &n
+}
+
+func (v *Proxy4Value) GetValuePtr() unsafe.Pointer {
+	return unsafe.Pointer(v)
+}
+
+func (k *Proxy4Value) Convert() *Proxy4Value {
+	n := *k
+	n.OrigDPort = common.Swab16(n.OrigDPort)
+	return &n
+}
+
+func (v *Proxy4Value) String() string {
+	return fmt.Sprintf("%s:%d", v.OrigDAddr.IP().String(), v.OrigDPort)
```

Comment:
I looked at this again and this is guaranteed to always be IPv4 so what is the point of using `JoinHostPort()`? It only required to convert the port to a string first.